hazelcast/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlanBuilder.java

```diff
@@ -279,10 +350,71 @@ public static Map<MemberInfo, int[]> getPartitionAssignment(
+                    nodeEngine.getLogger(ExecutionPlanBuilder.class).fine("Adding required member " + requiredMemberAddress +
+                            " to partition-pruned job members");
+                    // Extra members may get some partitions assigned later, especially for ALL_PARTITIONS_REQUIRED
+                    return new FixedCapacityIntArrayList(partitionCount);
+                });
+            });
+
+            // There is a special case of partition/member pruning: when DAG contains distributed-partitioned edge,
+            // we still want to apply member pruning, but we must redirect partitioned items to limited cluster subset.
+            // To do that, we assign all unassigned (also they are a non-required) partitions to all required members
+            // which are already was filtered by main assignment loop above.
+            if (pruningLevels.contains(ALL_PARTITIONS_REQUIRED) || !routingPartitions.isEmpty()) {
+                Set<Integer> partitionsToAssign = pruningLevels.contains(ALL_PARTITIONS_REQUIRED)
+                        ? new HashSet<>(range(0, partitionCount))
+                        : new HashSet<>(routingPartitions);
+                // do not assign duplicates, possible in both above cases
+                partitionsToAssign.removeAll(dataPartitions);
+
+                List<MemberInfo> requiredMembers = new ArrayList<>(partitionsForMember.keySet());
+                for (int partitionId : partitionsToAssign) {
+                    // Assign remaining partitions to one of the required members in round-robin fashion.
+                    // they will be only used for internal routing.
+                    //
+                    // The partition assignment is not balanced here, so not all members have the same number of partitions
+                    // especially for ALL_PARTITIONS_REQUIRED. This is not very important when there are only a few partitions
+                    // in dataPartitions, but can make some difference if there are many (eg. half of them).
+                    // This is not obvious where extra partitions should be assigned - maybe we should prefer members
+                    // that do not store the data for the job because they will be less loaded?
+                    var member = requiredMembers.get(memberIndex++ % requiredMembers.size());
+                    partitionsForMember.get(member).add(partitionId);
+                }
+            }
+        }
+
         Map<MemberInfo, int[]> partitionAssignment = new HashMap<>();
         for (Entry<MemberInfo, FixedCapacityIntArrayList> memberWithPartitions : partitionsForMember.entrySet()) {
-            partitionAssignment.put(memberWithPartitions.getKey(), memberWithPartitions.getValue().asArray());
+            int[] p = memberWithPartitions.getValue().asArray();
+            if (dataPartitions != null) {
+                // TODO: is the list of partitions expected to be sorted?
```

Comment:
@Fly-Style could you check if order is expected?

Reply:
Remove / adjust the comment then?

Reply:
I would rather document the requirement or order or lack of guarantee - this is very important to know if we want to make some optimisations (eg. binary search or removing redundant sorting). Sorting just-in-case is a bad approach.